log4j-api/src/main/java/org/apache/logging/log4j/util/ProviderUtil.java

```diff
@@ -63,13 +63,10 @@
     private static volatile ProviderUtil instance;
 
     private ProviderUtil() {
-        for (final ClassLoader classLoader : LoaderUtil.getClassLoaders()) {
-            try {
-                loadProviders(classLoader);
-            } catch (final Throwable ex) {
-                LOGGER.debug("Unable to retrieve provider from ClassLoader {}", classLoader, ex);
-            }
-        }
+        ServiceLoaderUtil.loadServices(Provider.class)
+                .filter(provider -> validVersion(provider.getVersions()) && !PROVIDERS.contains(provider))
+                .forEach(PROVIDERS::add);
```

Comment:
No need to check contains, add will return false and no-op if the element already exists